src/video/opengl.cpp

```diff
@@ -43,6 +43,49 @@
 #include "../safeguards.h"
 
 
+#if !defined(__APPLE__)
+typedef const GLubyte *(APIENTRYP PFNGLGETSTRINGPROC) (GLenum name);
+typedef void (APIENTRYP PFNGLGETINTEGERVPROC) (GLenum pname, GLint *params);
+typedef GLenum(APIENTRYP PFNGLGETERRORPROC) (void);
+typedef void (APIENTRYP PFNGLDISABLEPROC) (GLenum cap);
+typedef void (APIENTRYP PFNGLENABLEPROC) (GLenum cap);
+typedef void (APIENTRYP PFNGLVIEWPORTPROC) (GLint x, GLint y, GLsizei width, GLsizei height);
+typedef void (APIENTRYP PFNGLTEXIMAGE1DPROC) (GLenum target, GLint level, GLint internalformat, GLsizei width, GLint border, GLenum format, GLenum type, const GLvoid *pixels);
+typedef void (APIENTRYP PFNGLTEXIMAGE2DPROC) (GLenum target, GLint level, GLint internalformat, GLsizei width, GLsizei height, GLint border, GLenum format, GLenum type, const GLvoid *pixels);
+typedef void (APIENTRYP PFNGLTEXPARAMETERIPROC) (GLenum target, GLenum pname, GLint param);
+typedef void (APIENTRYP PFNGLTEXSUBIMAGE1DPROC) (GLenum target, GLint level, GLint xoffset, GLsizei width, GLenum format, GLenum type, const GLvoid *pixels);
+typedef void (APIENTRYP PFNGLTEXSUBIMAGE2DPROC) (GLenum target, GLint level, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLenum type, const GLvoid *pixels);
+typedef void (APIENTRYP PFNGLBINDTEXTUREPROC) (GLenum target, GLuint texture);
+typedef void (APIENTRYP PFNGLDELETETEXTURESPROC) (GLsizei n, const GLuint *textures);
+typedef void (APIENTRYP PFNGLGENTEXTURESPROC) (GLsizei n, GLuint *textures);
+typedef void (APIENTRYP PFNGLPIXELSTOREIPROC) (GLenum pname, GLint param);
+typedef void (APIENTRYP PFNGLCLEARPROC) (GLbitfield mask);
+typedef void (APIENTRYP PFNGLCLEARCOLORPROC) (GLfloat red, GLfloat green, GLfloat blue, GLfloat alpha);
+typedef void (APIENTRYP PFNGLBLENDFUNCPROC) (GLenum sfactor, GLenum dfactor);
+typedef void (APIENTRYP PFNGLDRAWARRAYSPROC) (GLenum mode, GLint first, GLsizei count);
+#endif
+
+static PFNGLGETSTRINGPROC _glGetString;
```

Comment:
To avoid defining possibly differing prototypes, you can use "decltype" to retrieve the prototype of a function:
```
using PFNGLGETSTRINGPROC = decltype(&glGetString);

static decltype(&glGetString) _glGetString;

#define GET_OGL_PROC_ADDRESS(func) reinterpret_cast<decltype(&func)>(GetOGLProcAddress(##func))

_glGetString = GET_OGL_PROC_ADDRESS(glGetString)
```

Reply:
Make a preprocessor horror with an include file that's just a list of OGL function names wrapped in a macro.

```c++
/// oglfunc.inc
F(glDisable)
F(glEnable)
F(glViewport)
...

/// opengl.cpp
#define F(func) static decltype(& func) _ # func;
#include "oglfunc.inc"
#undef F

static bool get_ogl_functions()
{
#define F(func) _ # func = reinterpret_cast<decltype(&func)>(GetOGLProcAddress(##func)); if (_ # func == nullptr) return false;
#include "oglfunc.h"
#undef F
  return true;
}
```